### PR TITLE
[15.0][MIG] payment: migration script

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Supersedes https://github.com/OCA/OpenUpgrade/pull/3304.

### CODE REVIEW

**Comment by sgheller** [comment]
Hello, I tried to test this PR and got this error
<details>
<summary> Error </summary>

```

2022-09-07 02:00:00,800 1 INFO devel OpenUpgrade: payment: loading 15.0.2.0/noupdate_changes.xml

2022-09-07 02:00:00,830 1 ERROR devel OpenUpgrade: payment: error in migration script /opt/odoo/custom/src/private/openupgrade_scripts/scripts/payment/15.0.2.0/post-migration.py: while parsing /opt/odoo/custom/src/private/openupgrade_scripts/scripts/payment/15.0.2.0/noupdate_changes.xml:37, somewhere inside
<record id="payment_acquirer_authorize" model="payment.acquirer">
    <field name="payment_icon_ids" eval="[(6, 0, [                    ref('payment.payment_icon_cc_maestro'),                    ref('payment.payment_icon_cc_mastercard'),
          ref('payment.payment_icon_cc_discover'),                    ref('payment.payment_icon_cc_diners_club_intl'),                    ref('payment.payment_icon_cc_jcb'),                    ref('payment.payment_icon_cc_visa'),                ])]"/>
    <field name="sequence"/>
  </record>

2022-09-07 02:00:00,830 1 ERROR devel OpenUpgrade: while parsing /opt/odoo/custom/src/private/openupgrade_scripts/scripts/payment/15.0.2.0/noupdate_changes.xml:37, somewhere inside
<record id="payment_acquirer_authorize" model="payment.acquirer">
    <field name="payment_icon_ids" eval="[(6, 0, [                    ref('payment.payment_icon_cc_maestro'),                    ref('payment.payment_icon_cc_mastercard'),
          ref('payment.payment_icon_cc_discover'),                    ref('payment.payment_icon_cc_diners_club_intl'),                    ref('payment.payment_icon_cc_jcb'),                    ref('payment.payment_icon_cc_visa'),                ])]"/>
    <field name="sequence"/>
  </record>
Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/tools/convert.py", line 680, in _tag_root
    f(rec)
  File "/opt/odoo/custom/src/odoo/odoo/tools/convert.py", line 575, in _tag_record
    f_val = int(f_val)
ValueError: invalid literal for int() with base 10: ''

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/openupgradelib/openupgrade.py", line 2042, in wrapped_function
    func(
  File "/opt/odoo/custom/src/private/openupgrade_scripts/scripts/payment/15.0.2.0/post-migration.py", line 61, in migrate
    openupgrade.load_data(env.cr, "payment", "15.0.2.0/noupdate_changes.xml")
  File "/usr/local/lib/python3.8/site-packages/openupgradelib/openupgrade.py", line 323, in load_data
    tools.convert_xml_import(cr, module_name, fp, idref, mode=mode)
  File "/opt/odoo/custom/src/odoo/odoo/tools/convert.py", line 811, in convert_xml_import
    obj.parse(doc.getroot())
  File "/opt/odoo/custom/src/odoo/odoo/tools/convert.py", line 731, in parse
    self._tag_root(de)
  File "/opt/odoo/custom/src/odoo/odoo/tools/convert.py", line 693, in _tag_root
    raise ParseError('while parsing %s:%s, somewhere inside\n%s' % (
odoo.tools.convert.ParseError: while parsing /opt/odoo/custom/src/private/openupgrade_scripts/scripts/payment/15.0.2.0/noupdate_changes.xml:37, somewhere inside
<record id="payment_acquirer_authorize" model="payment.acquirer">
    <field name="payment_icon_ids" eval="[(6, 0, [                    ref('payment.payment_icon_cc_maestro'),                    ref('payment.payment_icon_cc_mastercard'),
          ref('payment.payment_icon_cc_discover'),                    ref('payment.payment_icon_cc_diners_club_intl'),                    ref('payment.payment_icon_cc_jcb'),                    ref('payment.payment_icon_cc_visa'),                ])]"/>
    <field name="sequence"/>
  </record>

```

</details>

**Comment by bealdav** [comment]
Hi @MiquelRForgeFlow Finally account module migrated. What is the next step : rebase the PR ?
Thanks

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt [hunk 1]
```diff
@@ -0,0 +1,183 @@
+---Models in module 'payment'---
+new model payment.refund.wizard [transient]
+# NOTHING TO DO: new model: https://github.com/odoo/odoo/commit/e0233a1010618450cf53b997c50964bed2bb1b1a
+
+---Fields in module 'payment'---
+payment      / account.payment          / source_payment_id (many2one)  : NEW relation: account.payment, isrelated: related, stored
+# NOTHING TO DO: it's related to source_transaction_id and the new refund feature
+
+payment      / account.payment.method.line / payment_acquirer_id (many2one): NEW relation: payment.acquirer, isfunction: function, stored
+# NOTHING TO DO
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt]: You have to create records of this type for storing the journals of the `payment.acquirer`, following the inverse method:

https://github.com/odoo/odoo/blob/cdaca6d49281b6106db646ddfda2dd7e0d7c8388/addons/payment/models/payment_acquirer.py#L182

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+payment      / payment.acquirer         / _order                        : _order is now 'module_state, state desc, sequence, name' ('module_state, state, sequence, name')
+payment      / payment.acquirer         / allow_tokenization (boolean)  : NEW
+payment      / payment.acquirer         / check_validity (boolean)      : DEL
+payment      / payment.acquirer         / inline_form_view_id (many2one): NEW relation: ir.ui.view
+payment      / payment.acquirer         / journal_id (many2one)         : not stored anymore
+payment      / payment.acquirer         / journal_id (many2one)         : now a function
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt]: These 2 fields should go with the line about the model `account.payment.method.line`

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+payment      / payment.acquirer         / payment_flow (selection)      : DEL required, selection_keys: ['form', 's2s'], req_default: function
+# NOTHING TO DO
+
+payment      / payment.acquirer         / provider (selection)          : selection_keys is now '['none']' ('['manual']')
+# DONE: post-migration: converted value of provider (manual -> none)
+
+payment      / payment.acquirer         / qr_code (boolean)             : module is now 'payment_transfer' ('payment')
+payment      / payment.acquirer         / redirect_form_view_id (many2one): NEW relation: ir.ui.view
+payment      / payment.acquirer         / registration_view_template_id (many2one): DEL relation: ir.ui.view
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt]: This one has been renamed to `inline_form_view_id`

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+payment      / account.payment          / source_payment_id (many2one)  : NEW relation: account.payment, isrelated: related, stored
+# NOTHING TO DO: it's related to source_transaction_id and the new refund feature
+
+payment      / account.payment.method.line / payment_acquirer_id (many2one): NEW relation: payment.acquirer, isfunction: function, stored
+# NOTHING TO DO
+
+payment      / payment.acquirer         / _order                        : _order is now 'module_state, state desc, sequence, name' ('module_state, state, sequence, name')
+payment      / payment.acquirer         / allow_tokenization (boolean)  : NEW
+payment      / payment.acquirer         / check_validity (boolean)      : DEL
+payment      / payment.acquirer         / inline_form_view_id (many2one): NEW relation: ir.ui.view
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt]: This is the renamed field `registration_view_template_id`

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+payment      / payment.acquirer         / journal_id (many2one)         : not stored anymore
+payment      / payment.acquirer         / journal_id (many2one)         : now a function
+payment      / payment.acquirer         / payment_flow (selection)      : DEL required, selection_keys: ['form', 's2s'], req_default: function
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt]: Empty `inline_form_view_id` / `registration_view_template_id` when `payment_flow` = `form`

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+---Fields in module 'payment'---
+payment      / account.payment          / source_payment_id (many2one)  : NEW relation: account.payment, isrelated: related, stored
+# NOTHING TO DO: it's related to source_transaction_id and the new refund feature
+
+payment      / account.payment.method.line / payment_acquirer_id (many2one): NEW relation: payment.acquirer, isfunction: function, stored
+# NOTHING TO DO
+
+payment      / payment.acquirer         / _order                        : _order is now 'module_state, state desc, sequence, name' ('module_state, state, sequence, name')
+payment      / payment.acquirer         / allow_tokenization (boolean)  : NEW
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt]: Use the values from `save_token` to fill this one.

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+payment      / payment.acquirer         / check_validity (boolean)      : DEL
+payment      / payment.acquirer         / inline_form_view_id (many2one): NEW relation: ir.ui.view
+payment      / payment.acquirer         / journal_id (many2one)         : not stored anymore
+payment      / payment.acquirer         / journal_id (many2one)         : now a function
+payment      / payment.acquirer         / payment_flow (selection)      : DEL required, selection_keys: ['form', 's2s'], req_default: function
+# NOTHING TO DO
+
+payment      / payment.acquirer         / provider (selection)          : selection_keys is now '['none']' ('['manual']')
+# DONE: post-migration: converted value of provider (manual -> none)
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt:30]: You can do a copy of the column for backup purpose.

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+payment      / payment.acquirer         / save_token (selection)        : DEL selection_keys: ['always', 'ask', 'none']
+payment      / payment.acquirer         / support_authorization (boolean): NEW
+payment      / payment.acquirer         / support_fees_computation (boolean): NEW
+payment      / payment.acquirer         / support_refund (selection)    : NEW selection_keys: ['full_only', 'partial']
+payment      / payment.acquirer         / support_tokenization (boolean): NEW
+payment      / payment.acquirer         / view_template_id (many2one)   : DEL relation: ir.ui.view
+# NOTHING TO DO
+
+payment      / payment.icon             / _order                        : _order is now 'sequence, name' ('id')
+payment      / payment.icon             / image_payment_form (binary)   : now related
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt:42]: Not sure if this is processed automatically.

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+payment      / payment.token            / payment_ids (one2many)        : DEL relation: payment.transaction
+payment      / payment.token            / transaction_ids (one2many)    : NEW relation: payment.transaction
+# DONE: pre-migration: renamed field payment_ids -> transaction_ids
+
+payment      / payment.token            / provider (selection)          : previously in module payment_authorize
+# NOTHING TO DO
+
+payment      / payment.transaction      / callback_is_done (boolean)    : NEW
+payment      / payment.transaction      / company_id (many2one)         : NEW relation: res.company, isrelated: related, stored
+payment      / payment.transaction      / date (datetime)               : DEL
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt:60]:
```suggestion
payment      / payment.transaction      / date (datetime)               : DEL
# NOTHING TO DO: This date will be in the best case the same as write_date, and in the worst, the create_date, so nothing relevant.
```

File: openupgrade_scripts/scripts/payment/15.0.2.0/pre-migration.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+def fill_payment_transaction_partner_id(env):
+    # field it's required now
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE payment_transaction pt
+        SET partner_id = ap.partner_id
+        FROM account_payment ap
+        WHERE ap.partner_id IS NOT NULL
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/pre-migration.py]: Can it be the partner_id in the payment null? If so, the required may still not be fulfilled. If not, this condition is useless.

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt (hunk 1 above)
Comment (pedrobaeza) [reply openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt]: I see later that you have done it, so update properly these lines.

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+payment      / payment.transaction      / return_url (char)             : DEL
+payment      / payment.transaction      / source_transaction_id (many2one): NEW relation: payment.transaction
+# NOTHING TO DO
+
+payment      / payment.transaction      / payment_token_id (many2one)   : DEL relation: payment.token
+payment      / payment.transaction      / token_id (many2one)           : NEW relation: payment.token
+# DONE: pre-migration: renamed field payment_token_id -> token_id
+
+payment      / payment.transaction      / tokenize (boolean)            : NEW
```
Comment (pedrobaeza) [inline openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt:93]: Put it to True if `type` = `form_save`

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt (hunk 6 above)
Comment (MiquelRForgeFlow) [reply openupgrade_scripts/scripts/payment/15.0.2.0/upgrade_analysis_work.txt]: used "always" and "ask" values. ok?

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**